include/deal.II/base/mpi.templates.h

```diff
@@ -190,6 +190,142 @@ namespace Utilities
 
 
 
+    template<typename T>
+    std::map<unsigned int, T>
+    some_to_some(const MPI_Comm &comm,
+                 const std::map<unsigned int, T> &objects_to_send)
+    {
+#ifndef DEAL_II_WITH_MPI
+      (void)comm;
+      (void)objects_to_send;
+      Assert (false, ExcMessage ("The function some_to_some doesn't make"
+                                 "any sense if you do not have MPI enabled. "));
+#else
+      auto n_procs = dealii::Utilities::MPI::n_mpi_processes(comm);
+      auto my_proc = dealii::Utilities::MPI::this_mpi_process(comm);
+
+      std::vector<unsigned int> send_to(objects_to_send.size());
+      {
+        unsigned int i=0;
+        for (const auto &m: objects_to_send)
+          send_to[i++] = m.first;
+      }
+      AssertDimension(send_to.size(), objects_to_send.size());
+
+      const auto receive_from =
+        Utilities::MPI::compute_point_to_point_communication_pattern(comm, send_to);
+
+      // Sending buffers
+      std::vector<std::vector<char> > buffers_to_send(send_to.size());
+      std::vector<MPI_Request> buffer_send_requests(send_to.size());
+      {
+        unsigned int i = 0;
+        for (const auto &rank_obj : objects_to_send)
+          {
+            const auto &rank = rank_obj.first;
+            buffers_to_send[i] = Utilities::pack(rank_obj.second);
+            const int ierr = MPI_Isend(buffers_to_send[i].data(),
+                                       buffers_to_send[i].size(), MPI_CHAR,
+                                       rank, 21, comm, &buffer_send_requests[i]);
+            AssertThrowMPI(ierr);
+            ++i;
+          }
+      }
+
+      // Receiving buffers
+      std::map<unsigned int, T> received_objects;
+      {
+        unsigned int i = 0;
+        std::vector<char> buffer;
+
+        // We do this on a first come/first served basis
+        while (i<receive_from.size())
+          {
+            // Probe what's going on. Take data from the first available sender
+            MPI_Status status;
+            int ierr = MPI_Probe(MPI_ANY_SOURCE, 21, comm, &status);
+            AssertThrowMPI(ierr);
+
+            // Length of the message
+            int len;
+            ierr = MPI_Get_count(&status, MPI_CHAR, &len);
+            AssertThrowMPI(ierr);
+            buffer.resize(len);
+
+            // Source rank
+            const unsigned int rank = status.MPI_SOURCE;
+
+            // Actually receive the message
+            ierr = MPI_Recv(buffer.data(), len, MPI_CHAR,
+                            rank, 21, comm, MPI_STATUS_IGNORE);
+            AssertThrowMPI(ierr);
+
+            received_objects[rank] = Utilities::unpack<T>(buffer);
```

Comment:
before doing this, you can add an assert that `rank` is not contained in `received_objects` map.